c8run/internal/archive/archive.go

```diff
@@ -0,0 +1,289 @@
+package archive
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func DownloadFile(filepath string, url string) error {
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status: %s", resp.Status)
+	}
+
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func CreateTarGzArchive(files []string, buf io.Writer) error {
+	gw := gzip.NewWriter(buf)
+	defer gw.Close()
+	tw := tar.NewWriter(gw)
+	defer tw.Close()
+
+	for _, file := range files {
+		err := filepath.Walk(file, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() {
+				return addToArchive(tw, path)
+			} else {
+				// Add directory to the archive
+				header, err := tar.FileInfoHeader(info, path)
+				if err != nil {
+					return err
+				}
+				header.Name = path + "/"
+				if err := tw.WriteHeader(header); err != nil {
+					return err
+				}
+				return nil
+
+			}
+		})
+		if err != nil {
+			return err
```

Comment:
you might be interested to wrap the error and provide more information about which file had an error

Reply:
I added this context for all errors